jepsen: `PSQLException: The connection attempt failed` `Caused by: java.io.EOFException: null`

```
[12:29:44][/subcritical-skews+start-kill-2] ERROR [2017-11-21 12:29:43,623] main - jepsen.cli Oh jeez, I'm sorry, Jepsen broke. Here's why:
[12:29:44][/subcritical-skews+start-kill-2] org.postgresql.util.PSQLException: The connection attempt failed.
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.Driver$ConnectThread.getResult(Driver.java:358) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.Driver.connect(Driver.java:282) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.sql.DriverManager.getConnection(DriverManager.java:664) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.sql.DriverManager.getConnection(DriverManager.java:208) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] 	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:272) ~[classes/:na]
[12:29:44][/subcritical-skews+start-kill-2] 	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:193) ~[classes/:na]
[12:29:44][/subcritical-skews+start-kill-2] 	at jepsen.cockroach.client$client$open__1314$fn__1315.invoke(client.clj:90) ~[classes/:na]
[12:29:44][/subcritical-skews+start-kill-2] 	at clojure.core$binding_conveyor_fn$fn__4676.invoke(core.clj:1938) ~[clojure-1.8.0.jar:na]
[12:29:44][/subcritical-skews+start-kill-2] 	at clojure.lang.AFn.call(AFn.java:18) ~[clojure-1.8.0.jar:na]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] 	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_151]
[12:29:44][/subcritical-skews+start-kill-2] Caused by: java.io.EOFException: null
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.core.PGStream.receiveChar(PGStream.java:285) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:424) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:220) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:55) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:219) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.Driver.makeConnection(Driver.java:407) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.Driver.access$100(Driver.java:55) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	at org.postgresql.Driver$ConnectThread.run(Driver.java:317) ~[postgresql-9.4.1211.jar:9.4.1211]
[12:29:44][/subcritical-skews+start-kill-2] 	... 1 common frames omitted
```

From the server logs in #20593, we have a panic:

```
W171209 08:51:50.342206 413 storage/replica_range_lease.go:418  can't determine lease status due to node liveness error: node not in the liveness table
W171209 08:51:50.518491 102 storage/store.go:1289  [n1,s1,r7/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r7: replica (n1,s1):1 not lease holder; lease holder unknown
E171209 08:51:51.089407 472 util/log/crash_reporting.go:119  [n1,s1,r31/1:/{Table/50-Max}] a panic has occurred!
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1261959]

goroutine 472 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42076a2d0, 0x312fb00, 0xc425724380)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x136
panic(0x1e890a0, 0x30e36f0)
	/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*Progress).maybeUpdate(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/progress.go:116
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).appendEntry(0xc4207523c0, 0xc4256ee5a0, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:547 +0xd9
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).becomeLeader(0xc4207523c0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:648 +0x287
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).campaign(0xc4207523c0, 0x20d5e4a, 0x10)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:671 +0x668
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc4207523c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:783 +0x33a
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*RawNode).Campaign(0xc4256fc2c0, 0xc420496d80, 0x3)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/rawnode.go:140 +0x71
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc420161b00, 0x14fe9438e254d601, 0x21b2a38, 0x0, 0x200000002)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:594 +0x38e
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).maybeInitializeRaftGroup(0xc420161b00, 0x312fbc0, 0xc4256e93b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1849 +0x185
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send(0xc420161b00, 0x312fbc0, 0xc4256e93b0, 0x14fe943c314d23c9, 0x0, 0x0, 0x0, 0x1f, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1873 +0x136
github.com/cockroachdb/cockroach/pkg/storage.(*pendingLeaseRequest).requestLeaseAsync.func2(0x312fb00, 0xc425724380)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_range_lease.go:274 +0x577
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc42076a2d0, 0x312fb00, 0xc425724380, 0xc4207d0500, 0x35, 0x0, 0x0, 0xc4207b9b80)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:267 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:262 +0x133
```

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -30,7 +30,7 @@ import (
 	"github.com/coreos/etcd/raft/raftpb"
 	"github.com/google/btree"
 	"github.com/kr/pretty"
-	"github.com/opentracing/opentracing-go"
+	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 
@@ -4604,6 +4604,26 @@ func (r *Replica) processRaftCommand(
 			)
 			r.store.metrics.AddSSTableApplications.Inc(1)
 			raftCmd.ReplicatedEvalResult.AddSSTable = nil
+		}
+
+		if raftCmd.ReplicatedEvalResult.Split != nil {
+			// Splits require a new HardState to be written to the new RHS
+			// range (and this needs to be atomic with the main batch). This
+			// cannot be constructed at evaluation time because it differs
+			// on each replica (votes may have already been cast on the
+			// uninitialized replica). Transform the write batch to add the
+			// updated HardState.
+			// See https://github.com/cockroachdb/cockroach/issues/20629
+			//
+			// This is not the most efficient, but it only happens on splits,
+			// which are relatively infrequent and don't write much data.
+			tmpBatch := r.store.engine.NewBatch()
+			if err := tmpBatch.ApplyBatchRepr(writeBatch.Data, false); err != nil {
+				log.Fatal(ctx, err)
+			}
+			splitPreApply(ctx, r.store.cfg.Settings, tmpBatch, raftCmd.ReplicatedEvalResult.Split.SplitTrigger)
+			writeBatch.Data = tmpBatch.Repr()
+			tmpBatch.Close()
 		}
 
 		var delta enginepb.MVCCStats

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -28,7 +28,7 @@ import (
 	"github.com/coreos/etcd/raft"
 	"github.com/coreos/etcd/raft/raftpb"
 	"github.com/google/btree"
-	"github.com/opentracing/opentracing-go"
+	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"golang.org/x/time/rate"
@@ -1798,6 +1798,21 @@ func (s *Store) NewRangeDescriptor(
 	return desc, nil
 }
 
+// splitPreApply is called when the raft command is applied. Any
+// changes to the given ReadWriter will be written atomically with the
+// split commit.
+func splitPreApply(
+	ctx context.Context, st *cluster.Settings, eng engine.ReadWriter, split roachpb.SplitTrigger,
+) {
+	// Update the raft HardState with the new Commit value now that the
+	// replica is initialized (combining it with existing or default
+	// Term and Vote).
+	rsl := stateloader.Make(st, split.RightDesc.RangeID)
+	if err := rsl.SynthesizeRaftState(ctx, eng); err != nil {
+		log.Fatal(ctx, err)
+	}
+}
+
 // splitPostApply is the part of the split trigger which coordinates the actual
 // split with the Store. Requires that Replica.raftMu is held.
 //
@@ -1822,19 +1837,7 @@ func splitPostApply(
 		}
 	}
 
-	{
-		// Finish up the initialization of the RHS' RaftState now that we have
-		// committed the split Batch (which included the initialization of the
-		// ReplicaState). This will synthesize and persist the correct lastIndex and
-		// HardState.
-		rsl := stateloader.Make(r.store.cfg.Settings, split.RightDesc.RangeID)
-		if err := rsl.SynthesizeRaftState(ctx, r.store.Engine()); err != nil {
-			log.Fatal(ctx, err)
-		}
-	}
-
 	// Finish initialization of the RHS.
-
 	r.mu.Lock()
 	rightRng.mu.Lock()
 	// Copy the minLeaseProposedTS from the LHS.

